### PR TITLE
Start making TIR traces.

### PR DESCRIPTION
**Pull request by vext01**
This change starts building TIR traces by coping the relevant SIR statements into a TirTrace.

 * Guards are not yet implemented.
 * Tracing currently always yields invalid traces because we don't have a proper way to stop tracing at a merge point: we always hit a function for which we have no SIR, meaning that the trace is invalid. There is an API compromise so that we can test *something* in the interim.

No companion PR for this one :)

### CODE REVIEW

File: yktrace/src/tir.rs [hunk 1]
```diff
@@ -36,18 +36,96 @@ lazy_static! {
...
+    /// *something* to test in the interim.
+    pub(crate) fn new(trace: &'_ dyn SirTrace) -> Result<Self, Self> {
+        let mut ops = Vec::new();
+        let num_locs = trace.len();
+
+        for blk_idx in 0..num_locs {
+            let loc = trace.loc(blk_idx);
+            let body = match SIR_MAP.get(&DefId::from_sir_loc(loc)) {
+                Some(b) => b,
+                _ => return Err(Self { ops })
```
Comment (ltratt) [inline yktrace/src/tir.rs]: Can this be anything other than `None`?

File: yktrace/src/tir.rs [hunk 2]
```diff
@@ -36,18 +36,96 @@ lazy_static! {
...
+            };
+
+            let shadow_bb_idx_usize = usize::try_from(loc.bb_idx()).unwrap();
```
Comment (ltratt) [inline yktrace/src/tir.rs]: What type does `bb_idx` return?

File: yktrace/src/tir.rs (hunk 1 above)
Comment (vext01) [reply yktrace/src/tir.rs]: It returns `Some` if there is an entry in the SIR map, but that can't be what you are asking?

File: yktrace/src/tir.rs [hunk 3]
```diff
@@ -36,18 +36,96 @@ lazy_static! {
...
+    }
+
+    use crate::{start_tracing, TracingKind, TirTrace};
+    use test::black_box;
+
+    #[test]
+    fn nonempty_tir_trace() {
+        let tracer = start_tracing(Some(TracingKind::SoftwareTracing));
+        let res = black_box(work(black_box(3), black_box(13)));
+        println!("THis is a teds");
```
Comment (ltratt) [inline yktrace/src/tir.rs]: Huh?

File: yktrace/src/tir.rs (hunk 1 above)
Comment (ltratt) [reply yktrace/src/tir.rs]: Let me rephrase: let's use `None` instead of `_` if that's the only thing the `_` can match.

File: yktrace/src/tir.rs (hunk 2 above)
Comment (vext01) [reply yktrace/src/tir.rs]: https://github.com/softdevteam/ykrustc/blob/4bcbb5e45755cdd81d76cadd6917420981bdbeb2/src/libcore/yk_swt.rs#L35

File: yktrace/src/tir.rs (hunk 3 above)
Comment (vext01) [reply yktrace/src/tir.rs]: LOL! That's debugging code!

Comment (vext01) [reply yktrace/src/tir.rs]: I think we only need to black box the args too.

File: yktrace/src/tir.rs (hunk 1 above)
Comment (vext01) [reply yktrace/src/tir.rs]: Why? The type system will ensure we are exhaustively matching.

Comment (ltratt) [reply yktrace/src/tir.rs]: That statement makes little sense when you use the wildcard `_` operator: by definiton you've exhaustively matched everything.

Comment (vext01) [reply yktrace/src/tir.rs]: So think you are asking me to change `_` to `None`. We can, but it doesn't seem necessary. I won't argue over it :P

**Comment by vext01** [comment]
Oddly, tracing is not erroring out any more. It was yesterday... I'll make some changes to my tests too.

File: yktrace/src/tir.rs (hunk 1 above)
Comment (ltratt) [reply yktrace/src/tir.rs]: Yes, we should use `None`. In general, `_` is something that one should be very wary of in production code (though in tests I'm much less worried about it).

Comment (vext01) [reply yktrace/src/tir.rs]: As you wish, but I'd have though it was obvious that `_` can only be `None` since the previous match was `Some`.

Comment (ltratt) [reply yktrace/src/tir.rs]: What happens if the type you return changes in some way? The problem with `_` is that it allows you to sidestep most of the goodies of the type system, and that tends to bite you on the rear end years down the line during refactoring...

Comment (vext01) [reply yktrace/src/tir.rs]: OK, I take your point.

**Comment by vext01** [comment]
Give this a shot.

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
Splat and reword commit msg.

**Comment by vext01** [comment]
```
+ cargo fmt --all -- --check
Diff in /home/buildbot-worker6/worker6/buildbot-build-script/build/ykpack/src/types.rs at line 9:
 
 //! Types for the Yorick intermediate language.
 
+use core::yk_swt::SirLoc;
 use serde::{Deserialize, Serialize};
 use std::fmt::{self, Display};
-use core::yk_swt::SirLoc;
 
 pub type CrateHash = u64;
 pub type DefIndex = u32;
Diff in /home/buildbot-worker6/worker6/buildbot-build-script/build/yktrace/src/tir.rs at line 101:
 /// A TIR operation. A collection of these makes a TIR trace.
 #[derive(Debug)]
 pub enum TirOp {
-    Statement(Statement)
-    // FIXME guards
+    Statement(Statement) // FIXME guards
 }
```

Odd. I literally just fmted. Must have screwed up. OK to squash?

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat.

**Comment by vext01** [comment]
```
+ cargo fmt --all -- --check
Diff in /home/buildbot-worker3/worker3/buildbot-build-script/build/ykpack/src/types.rs at line 9:
 
 //! Types for the Yorick intermediate language.
 
+use core::yk_swt::SirLoc;
 use serde::{Deserialize, Serialize};
 use std::fmt::{self, Display};
-use core::yk_swt::SirLoc;
 
 pub type CrateHash = u64;
 pub type DefIndex = u32;
```

Something fishy here.

```
$ cargo +nightly fmt -- --check
$
```

**Comment by vext01** [comment]
Got it. It's the `--all` arg!

Ok to squash *again*?

**Comment by ltratt** [comment]
OK, please squash. Fingers crossed!

**Comment by vext01** [comment]
splat
